spec/app/ics-070-shared-queue/README.md

```diff
@@ -0,0 +1,485 @@
+- Dequeue and peak operations are invoked solely by the baby blockchain.
+
+Lastly, our implementation satisfies the **sequential consistency** correctness criterium:
+The result of any execution is the same as if the operations of all the processes were executed in some sequential order, and the operations of each individual process appear in this sequence in the order specified by its protocol.
+
+#### Causally Consistent FIFO Queue
+
+We denote the second version of our FIFO queue by Causally Consistent FIFO Queue (CCQ), which exposes the following interface:
+- *enqueue(x)* operation: Operation that enqueues the element *x* to the queue.
+- *dequeue()* operation: Operation that dequeues the first element of the queue and returns that element.
+- *read()* operation: Operation that returns the "content" of the entire queue.
+
+We introduce the following constraints:
+- Enqueue operations are invoked solely by the parent blockchain.
+- Dequeue operations are invoked solely by the baby blockchain.
+
+Note the difference between causally and sequentially consistent FIFO queue in their interface.
+The SCQ exposes the peak and contains operations (that are invoked solely by the baby and parent blockchain, respectivelly), whereas CCQ exposes the read operation (that can be invoked by both blockchains).
+Note that read operation of the CCQ could be used to implement the peak and contains operations (i.e., the read operation could be reduced to the peak and contains operations).
+However, we show that the approach used for implementing SCQ, which is sequentially consistent, **cannot** be used for implementing sequentially consistent queue with the read operation.
+
+Thus, CCQ satisfies the **causal consistency** correctness criterium.
+Let us first define the causal precedence relation:
+Consider operations A and B.
+We say that A *causally precedes* B (we write "A --> B") if and only if:
+1) A and B are invoked by the same blockchain and A is invoked before B, or
+2) A is a write operation and the blockchain that invokes B has observed A before invoking B, or
+3) There exists an operation C such that A --> C and C --> B.  
+Moreover, operations A and B are *concurrent* if neither causally precedes the other.
+
+Finally, we are able to define the causal consistency correctness criterium.
+Operations that are related by the causal precedence relation are observed by all parties (i.e., blockchains) in their causal precedence order.
+Note that the definition allows concurrent operations to be seen in different order by different blockchains.
+
+### Closer Look at the IBC Channels
+
+An IBC channel assumes two parties (the respective blockchains) involved in the communication. However, it also assumes a relayer which handles message transmissions between the two blockchains. The relayer carries a central responsibility in ensuring communication between the two parties through the channel.
```

Comment:
Hmm, the relayer plays a role, yes, but how would you avoid that with a different design? I'm confused.

Reply:
Sorry, I do not understand your question.

Reply:
I think we need to be more specific. IBC requires the existence of at least one correct relayer for liveness, but that's a pretty weak requirement, the relayer has nothing to do with correctness and only one correct relayer needs to exist.